x/ibc/04-channel/keeper/packet.go

```diff
@@ -0,0 +1,307 @@
+	if !found {
+		return errors.New("consensus state not found") // TODO: sdk.Error
+	}
+
+	if consensusState.GetHeight() >= packet.TimeoutHeight() {
+		return types.ErrPacketTimeout(k.codespace)
+	}
+
+	nextSequenceSend, found := k.GetNextSequenceSend(ctx, packet.SourcePort(), packet.SourceChannel())
+	if !found {
+		return types.ErrSequenceNotFound(k.codespace, "send")
+	}
+
+	if packet.Sequence() != nextSequenceSend {
+		return types.ErrInvalidPacketSequence(k.codespace)
+	}
+
+	nextSequenceSend++
+	k.SetNextSequenceSend(ctx, packet.SourcePort(), packet.SourceChannel(), nextSequenceSend)
+	k.SetPacketCommitment(ctx, packet.SourcePort(), packet.SourceChannel(), packet.Sequence(), packet.Data()) // TODO: hash packet data
+
+	return nil
+}
+
+// RecvPacket is called by a module in order to receive & process an IBC packet
+// sent on the corresponding channel end on the counterparty chain.
+func (k Keeper) RecvPacket(
+	ctx sdk.Context,
+	packet exported.PacketI,
+	proof ics23.Proof,
+	proofHeight uint64,
+	acknowledgement []byte,
+) (exported.PacketI, error) {
+
+	channel, found := k.GetChannel(ctx, packet.SourcePort(), packet.SourceChannel())
+	if !found {
+		return nil, types.ErrChannelNotFound(k.codespace)
+	}
+
+	if channel.State != types.OPEN {
+		return nil, errors.New("channel not open") // TODO: sdk.Error
+	}
+
+	_, found = k.GetChannelCapability(ctx, packet.SourcePort(), packet.SourceChannel())
+	if !found {
+		return nil, types.ErrChannelCapabilityNotFound(k.codespace)
+	}
+
+	// if !AuthenticateCapabilityKey(capabilityKey) {
+	//  return errors.New("invalid capability key") // TODO: sdk.Error
+	// }
+
+	// packet must come from the channel's counterparty
+	if packet.SourcePort() != channel.Counterparty.PortID {
+		return nil, errors.New("invalid packet source port")
+	}
+
+	if packet.SourceChannel() != channel.Counterparty.ChannelID {
+		return nil, errors.New("invalid packet source channel")
+	}
+
+	connection, found := k.connectionKeeper.GetConnection(ctx, channel.ConnectionHops[0])
+	if !found {
+		return nil, ics03types.ErrConnectionNotFound(k.codespace)
+	}
+
+	if connection.State != ics03types.OPEN {
+		return nil, errors.New("connection is not open") // TODO: ics03 sdk.Error
+	}
+
+	if uint64(ctx.BlockHeight()) >= packet.TimeoutHeight() {
+		return nil, types.ErrPacketTimeout(k.codespace)
+	}
+
+	if !k.connectionKeeper.VerifyMembership(
+		ctx, connection, proofHeight, proof,
+		types.PacketCommitmentPath(packet.SourcePort(), packet.SourceChannel(), packet.Sequence()),
+		packet.Data(), // TODO: hash data
+	) {
+		return nil, errors.New("couldn't verify counterparty packet commitment")
+	}
+
+	if len(acknowledgement) > 0 || channel.Ordering == types.UNORDERED {
+		k.SetPacketAcknowledgement(
+			ctx, packet.DestPort(), packet.DestChannel(), packet.Sequence(),
+			acknowledgement, // TODO: hash ACK
+		)
+	}
+
+	if channel.Ordering == types.ORDERED {
```

Comment:
This should probably be moved up to before the membership verification to reject the tx earlier if the sequence is wrong.